[Lo Yong Zhe] iP

Reallyeasy1 opened the pull request:
# Citadel

> "A fortress is not merely a place of defense, but a symbol of strength, endurance, and resilience."

Citadel is a task management assistant that frees your mind from the burden of remembering things you need to do. It's:

- **text-based**
- **easy to learn**
- **FAST** – _SUPER FAST_ to use

All you need to do is:

1. Download it from [here](https://github.com/Reallyeasy1/ip).
2. Double-click it.
3. Add your tasks.
4. Let it manage your tasks for you 😉.

And it's **FREE**!

## Features:

- [x] Managing tasks
- [x] Managing deadlines
- [ ] Reminders (coming soon)
- [ ] Do automated trading system (coming soon)

If you are a Java programmer, you can use Citadel to practice Java too. Here's the main method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

Reallyeasy1 committed:
Rename, Greet and Exit

Reallyeasy1 committed:
Removed Print Duke

Reallyeasy1 committed:
Rename, Greet and Exit

Reallyeasy1 committed:
Added Echo to Chatbot

Reallyeasy1 committed:
Added List Data Structure and Add Texts features to Chatbot

Reallyeasy1 committed:
Added Task Class, and added marking and unmarking of tasks

Reallyeasy1 committed:
Added Deadline, Event and ToDo classes

Reallyeasy1 committed:
Edited Event and Deadline class and added the control flow for handling different task types

Reallyeasy1 committed:
Refactor code

Reallyeasy1 committed:
Added CitadelException, CitadelInvalidArgException, CitadelInvalidCommandException, CitadelTaskNoInput exceptions and exception handling

Reallyeasy1 committed:
Amended code to resolve code warnings

Reallyeasy1 committed:
Added Delete Feature

Reallyeasy1 committed:
Added Enums for the commands

Reallyeasy1 committed:
Add Database Save Feature to Chatbot

Reallyeasy1 committed:
Merge pull request #1 from Reallyeasy1/branch-level-7

Add Database Save Feature to Chatbot

Reallyeasy1 committed:
Use LocalDateTime for event and deadline

Reallyeasy1 committed:
Handle wrong DateTime format error

Reallyeasy1 committed:
Add check for if start time is before end time for event task

Reallyeasy1 committed:
Merge pull request #2 from Reallyeasy1/branch-level-8

Branch level 8

Reallyeasy1 committed:
Add different printing formats for datetime

Reallyeasy1 committed:
Merge pull request #3 from Reallyeasy1/branch-level-8

Add different printing formats for datetime

Reallyeasy1 committed:
Changed Database to Storage

Reallyeasy1 committed:
Change ArrayList to TaskList to store Tasks

Reallyeasy1 committed:
Fix bugs regarding TaskList

Reallyeasy1 committed:
Separate out the handle functions as classes and create TextUi class

Reallyeasy1 committed:
Allow TextUI to listen for inputs and display texts for all commands

Reallyeasy1 committed:
Add Parser class

Reallyeasy1 committed:
Move all files/ packages into Citadel Package

Reallyeasy1 committed:
Merge branch 'master' into add-gradle-support

Reallyeasy1 committed:
Added JUnit Test

Reallyeasy1 committed:
Merge pull request #4 from Reallyeasy1/add-gradle-support

Add gradle support and JUnit Tests

Reallyeasy1 committed:
Refactor naming of packages

Reallyeasy1 committed:
Change Class names into Pascal Case

Reallyeasy1 committed:
Rename CONST to SCREAMING_SNAKE_CASE

Reallyeasy1 committed:
Improve code layout by splitting long lines

Reallyeasy1 committed:
Improve code layout by splitting long lines part 2

Reallyeasy1 committed:
Use switch statement to handle if else statement

Reallyeasy1 committed:
Format Import structure

Reallyeasy1 committed:
Reformat variables and spacing

Reallyeasy1 committed:
Merge pull request #6 from Reallyeasy1/A-CodingStandard

Adhere Coding Standard

Reallyeasy1 committed:
Add JavaDocs

Reallyeasy1 committed:
Partially done with printFind

Reallyeasy1 committed:
Add Find

Reallyeasy1 committed:
Merge conflicts

Reallyeasy1 committed:
Resolve merge conflicts

Reallyeasy1 committed:
Add JavaDocs to Citadel and TextUI

Reallyeasy1 committed:
Add :D to introduction

Reallyeasy1 committed:
Add checkStyle to the project

Reallyeasy1 committed:
Add JavaFX tutorial/ base layout

Reallyeasy1 committed:
Add picture to JavaFX tutorial

Reallyeasy1 committed:
Add handleInput function for Citadel

Reallyeasy1 committed:
Refactor all Command classes to return String instead of void

Reallyeasy1 committed:
Refactor getResponse function to add base handleInput feature to frontend

Reallyeasy1 committed:
Refactor code to adhere to CheckStyle

Reallyeasy1 committed:
Refactor code to adhere to CheckStyle part 2

Reallyeasy1 committed:
Add intro message to GUI

Reallyeasy1 committed:
Allow Bye to close GUI Application

Reallyeasy1 committed:
Add loadDatabase and saveDatabase to GUI

gladwintan commented on src/main/java/citadel/Citadel.java:
```diff
@@ -0,0 +1,146 @@
+    /** The storage system used to persist tasks between sessions. */
+    private static final Storage db = new Storage("data/citadel");
+
+    /** The user interface used to interact with the user. */
```
Could this be replaced with normal comments instead of Javadoc comments?

Reallyeasy1 replied:
Good catch

gladwintan commented on src/main/java/citadel/Storage.java:
```diff
@@ -0,0 +1,183 @@
+import java.util.Scanner;
+import java.time.LocalDateTime;
+import java.time.format.DateTimeFormatter;
+
```
Good segmentation of different imports

Reallyeasy1 replied:
Thanks!

gladwintan commented on src/main/java/citadel/Storage.java:
```diff
@@ -0,0 +1,183 @@
+
+    /**
+     * Retrieves the list of tasks from the storage file.
+     * <p>
```
Any reason for using \<p>?

gladwintan commented on src/main/java/citadel/Task/Event.java:
```diff
@@ -0,0 +1,71 @@
+     */
+    @Override
+    public String printTask() {
+        return "[E]" + super.toString() + " (from: " + this.from
```
Should 'this.from' be changed to 'from' instead since there is no conflicting variable in the function?

gladwintan commented on src/main/java/citadel/Task/Task.java:
```diff
@@ -0,0 +1,70 @@
+     */
+    @Override
+    public String toString() {
+        return "[" + this.getStatusIcon() + "] " + this.description;
```
Should this.getStatusIcon() be changed to getStatusIcon() ?

gladwintan commented on src/main/java/citadel/ui/TextUI.java:
```diff
@@ -0,0 +1,191 @@
+public class TextUI {
+
+    /** The name of the Citadel application. */
+    private static final String NAME = "citadel";
```
Good naming of constants

Reallyeasy1 replied:
Thanks!

gladwintan reviewed (commented):
Overall, I found your code easy to read without any major problems.

josh1248 commented on src/main/java/citadel/Citadel.java:
```diff
@@ -0,0 +1,146 @@
+        case FIND:
+            return new FindTask(input, items).run();
+
+        default:
```
good catch of the default branch.

josh1248 commented on src/main/java/citadel/Citadel.java:
```diff
@@ -0,0 +1,146 @@
+        loadDatabase();  // Load tasks from storage
+        ui.printTasks(items);
+
+        while (true) {
```
this is very similar to your previous 'getResponse' method - do consider if code re-use here would help to reduce repeated code.

josh1248 commented on src/main/java/citadel/Main.java:
```diff
@@ -0,0 +1,33 @@
+import javafx.stage.Stage;
+
+/**
+ * A GUI for Duke using FXML.
```
do update your comments to your new chatbot name.

josh1248 commented on src/main/java/citadel/Storage.java:
```diff
@@ -0,0 +1,183 @@
+        return t;
+    }
+
+    /**
```
good javadocs!

josh1248 commented on src/main/java/citadel/Storage.java:
```diff
@@ -0,0 +1,183 @@
+        if (todo.isEmpty()) {
+            throw new CitadelTaskNoInput();
+        }
+        t = new ToDo(todo);
```
t is not very necessary here - could return directly since "t" does not provide extra description of your todo.

josh1248 commented on src/main/java/citadel/Task/Task.java:
```diff
@@ -0,0 +1,70 @@
+    protected String description;
+
+    /** The status of the task, indicating whether it is done. */
+    protected boolean isDone;
```
good naming of a boolean variable

josh1248 commented on src/main/java/citadel/ui/TextUI.java:
```diff
@@ -0,0 +1,191 @@
+     * @param t The {@link Task} that was deleted.
+     */
+    public static String printDelete(TaskList tasks, Task t) {
+        String removeSuccess = "Noted. I've removed this task:";
```
good variable name length - descriptive enough but not verbose.

josh1248 commented on src/main/resources/images/DaUser.png:
nice

Reallyeasy1 committed:
Add assertions

- Added assert statements in Parser.java to check for null and empty inputs when parsing commands.
- Implemented assert checks in Citadel.java to validate input strings, parsed commands, and task list existence before database operations.
- Added assertions in TextUI.java to ensure the task list and task representations are not null before printing.
- Added assert statements in Task.java to prevent duplicate marking or unmarking of tasks.
- Improved overall error handling and program robustness by validating critical variables before executing logic.

These changes help catch potential issues early, ensuring that the program only operates on valid data.

Reallyeasy1 committed:
Merge pull request #10 from Reallyeasy1/branch-A-Assertions

Add assertions

Reallyeasy1 committed:
Refactor TextUI/Command subclasses

- Refactored `TextUI` class:
  - Extracted common printing logic into a reusable `printMessage` method to reduce code duplication.
  - Added method `printTaskStatus` to consolidate task status (mark/unmark) printing logic.
  - Used `String.format` for better readability and efficiency when constructing strings.
  - Replaced magic strings and numbers with named constants where appropriate.
  - Applied guard clauses in methods to simplify error checking and reduce nesting.

  - Refactored Command subclasses
  - Add code comments for HandleDeadline, HandleTodo, MarkTask and UnmarkTask

Reallyeasy1 committed:
Refactor code part 2

Refactor Storage class: Add code comments explaining the purpose and rationale

- Added detailed comments in the Storage class focusing on explaining WHAT the code is doing and WHY it's structured in a particular way.
- Clarified the purpose of key methods such as `getTasks()`, `saveData()`, `createDeadline()`, `createEvent()`, and `createTodo()` by specifying their roles in managing tasks and interacting with the storage file.
- Provided explanations for input validation to ensure that task descriptions and deadlines are properly formatted and complete.
- Explained the rationale behind parsing string data into LocalDateTime objects for accurate task scheduling.
- Enhanced readability by adding comments where decisions were made to ensure data integrity, such as checking for empty strings or incomplete input.
- This refactor improves the code's maintainability and clarity, ensuring future developers can easily understand the purpose and logic behind each method without needing to infer the reason from the code alone.

Refactor Citadel class
Renamed processCommand() to handleInput() for consistency.
Improved error handling in getResponse() and the main loop, ensuring exceptions like CitadelException and DateTimeParseException are properly managed.
Ensured command processing is consistent and reliable, with explicit checks for valid input and commands.
Maintained the main application flow with better input validation and task saving.

Others:
Add code comments for deleteTask, HandleEvent TaskList, and Storage

Reallyeasy1 committed:
Merge branch 'master' into A-CodeQuality

Reallyeasy1 committed:
Merge pull request #11 from Reallyeasy1/A-CodeQuality

A code quality

Reallyeasy1 committed:
Fix Citadel bug

Reallyeasy1 committed:
Add tags ArrayList and functionality to Task

Reallyeasy1 committed:
Add Gradle jar file

Reallyeasy1 committed:
Add save feature to tags in tasks

Reallyeasy1 committed:
Refactor TodoTest and imports for Citadel

Reallyeasy1 committed:
Amend Event and Deadline dateparse error

Reallyeasy1 committed:
Add Ui picture

Reallyeasy1 committed:
Add README.md

Reallyeasy1 committed:
Change list to List

Reallyeasy1 committed:
Add Bye to README.md

Reallyeasy1 committed:
Format README.md

Reallyeasy1 committed:
Reformat README.md

Reallyeasy1 committed:
Reformat README.md part 2

Reallyeasy1 committed:
Reformat README.md part 4

Reallyeasy1 committed:
Refactor code on code quality

Reallyeasy1 committed:
Update README.md

Reallyeasy1 committed:
Update README.md

Reallyeasy1 committed:
Update README.md

Reallyeasy1 committed:
Update README.md

Reallyeasy1 committed:
Update README.md
